src/file.cpp

```diff
@@ -2011,11 +2015,7 @@ typedef struct _FILE_ALLOCATED_RANGE_BUFFER {
 		static SetFileValidData_t const pSetFileValidData = (SetFileValidData_t)
 				GetProcAddress(GetModuleHandleA("kernel32"), "SetFileValidData");
 
-		if (pSetFileValidData == NULL) return;
-
-		static bool const has_manage_volume_privs = get_manage_volume_privs();
-
-		if (has_manage_volume_privs)
+		if (pSetFileValidData && get_manage_volume_privs())
```

Comment:
I don't think you have to call this function here. It's already been called once before you get here. The approach in master was to ignore the result from it, and try the call to `SetFileValidData()` anyway, and let it fail, so I don't think there's any need for extra checking here either

Reply:
Why should we ask system to do something that is known to fail? This call is cheap because we already know result. Why not using that result to avoid call that would fail?

Reply:
because I'm not confident there is a 1:1 relationship between the result and whether the call will succeed. I have not found very clear documentation around this.

Reply:
I've added link to related doc in the comment. See `open_file()`.

Reply:
Here is also a good reading on this from Raymond Chen:
https://blogs.msdn.microsoft.com/oldnewthing/20160603-00/?p=93565

Reply:
that example code doesn't call `AdjustTokenPrivileges()` nor does it suggest that the `SetFileValidData()` can only succeed if the call to `AdjustTokenPrivileges()` succeeds. For instance, I would expect that `AdjustTokenPrivileges()` might fail for some reason other than lack of privileges.

Reply:
Please pay attention to [SetPrivelege](https://msdn.microsoft.com/library/windows/desktop/aa446619(v=vs.85).aspx), it hides a hyperlink.

Reply:
Possible corner cases also should include this, enabling required privilege once doesn't make sense since other piece of code could simply disable that privilege at any time.